### PR TITLE
Configurable TIM Signing for RSUs/SDX, Code Refactoring & UDP/Log Ingestion Updates

### PR DESCRIPTION
**Pull request by dmccoystephenson**
# PR Details
## Description
The project has undergone a number of significant updates. Overall, these changes collectively aim to improve functionality, documentation clarity, and system performance.

#### Dependencies Updates
- Kafka updated to version 3.6.1
- Zookeeper removed from all setup scripts and documentation
- Several Maven plugins upgraded
- Kafka init container introduced

#### Documentation Revisions
- Extensive accuracy improvements across multiple guides
- Added Submodule Compatibility Guide
- Updated data flow diagrams

#### Feature Enhancements
- Configurable signing capabilities for RSUs and SDX introduced
- Support for configurable default SNMP protocol
- Modified SnmpSession class to handle msgRepeatOptions based on signing context
- Improved reporting of failures to retrieve authoritative engine IDs
- Added ability to set up a MongoDB database with Kafka connect instance to sink specified topics
- Updated abstract UDP receiver class to include more functionality
- Log ingestion now supports more than MAP/TIM message types
- Added UperUtil class containing message type start flags and 1609.2/1609.3 trim functionality

#### Code Refactoring and Improvements
- Enhanced Asn1EncodedDataRouter class to strip 1609.2 headers
- Corrected handling of timestamps and datetime values for precision
- Improved log file handling
- Enhanced payload parsing
- Refactored receiver classes for robustness and maintainability

## Related Issue
No related GitHub issue.

## Motivation and Context
Updating dependencies such as Kafka and Maven plugins ensures compatibility with the latest features and security patches, while removing Zookeeper simplifies the development environment. Introducing configurable signing for RSUs and the SDX, along with a configurable default SNMP protocol, adds flexibility and improves security. Enhancements to the SNMP protocol further strengthen system reliability.

Extensive documentation revisions, including the addition of a Submodule Compatibility Guide and updated data flow diagrams, improve clarity and accuracy for developers and users. Code refactoring efforts, such as improving timestamp handling and restructuring receiver classes, streamline the codebase, enhancing readability and maintainability.

## How Has This Been Tested?
Unit tests have been verified to pass with these changes. These changes have also been tested in CDOT's GCP environment.

## Types of changes
- [x] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:
- [ ] I have added any new packages to the sonar-scanner.properties file
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[ODE Contributing Guide](https://github.com/usdot-jpo-ode/jpo-ode/blob/bugfix/Pull_request_template/docs/contributing_guide.md) 
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: jpo-ode-svcs/src/main/resources/application.properties [hunk 1]
```diff
@@ -19,18 +19,9 @@ spring.http.multipart.max-request-size=1MB
 #ode.uploadLocationBsm = bsm
 #ode.uploadLocationMessageFrame = messageframe
 
-#USDOT Situation Data Clearinghouse (SDC)/ Situation Data Warehouse (SDW), a.k.a Data Distribution System (DDS) Properties
+#USDOT Situation Data Clearinghouse (SDC)/ Situation Data Warehouse (SDW) Properties
```
Comment (dan-du-car) [inline jpo-ode-svcs/src/main/resources/application.properties:22]: SDW or SDX?

File: quickstart-compose.yml [hunk 2]
```diff
@@ -2,35 +2,33 @@
...
-                                     # file isn't full to ensure that retention can delete or compact old data.
-    depends_on:
-      - zookeeper
-    volumes:
-      - ${DOCKER_SHARED_VOLUME_WINDOWS}/var/run/docker.sock:/var/run/docker.sock
-
+      KAFKA_CREATE_TOPICS: "topic.OdeBsmPojo:1:1,topic.OdeSpatTxPojo:1:1,topic.OdeSpatPojo:1:1,topic.OdeSpatJson:1:1,topic.FilteredOdeSpatJson:1:1,topic.OdeSpatRxJson:1:1,topic.OdeSpatRxPojo:1:1,topic.OdeBsmJson:1:1,topic.FilteredOdeBsmJson:1:1,topic.OdeTimJson:1:1,topic.OdeTimBroadcastJson:1:1,topic.J2735TimBroadcastJson:1:1,topic.OdeDriverAlertJson:1:1,topic.Asn1DecoderInput:1:1,topic.Asn1DecoderOutput:1:1,topic.Asn1EncoderInput:1:1,topic.Asn1EncoderOutput:1:1,topic.SDWDepositorInput:1:1,topic.OdeTIMCertExpirationTimeJson:1:1,topic.OdeRawEncodedBSMJson:1:1,topic.OdeRawEncodedSPATJson:1:1,topic.OdeRawEncodedTIMJson:1:1,topic.OdeRawEncodedMAPJson:1:1,topic.OdeMapTxPojo:1:1,topic.OdeMapJson:1:1,topic.OdeRawEncodedSSMJson:1:1,topic.OdeSsmPojo:1:1,topic.OdeSsmJson:1:1,topic.OdeRawEncodedSRMJson:1:1,topic.OdeSrmTxPojo:1:1,topic.OdeSrmJson:1:1,topic.OdeRawEncodedPSMJson:1:1,topic.OdePsmTxPojo:1:1,topic.OdePsmJson:1:1"
```
Comment (dan-du-car) [inline quickstart-compose.yml:24]: The same as above, is this CREATE TOPICS replaced by the kafka_init.sh script?

File: docs/contributing_guide.md [hunk 3]
```diff
@@ -50,9 +48,16 @@ All pull requests will be reviewed by the JPO-ODE team. The team member will eit
 	5.	Update the documentation.
 		- User QA procedures are documented within the Github Wiki
 		- Architecture and user guide documentation should be included in the word document under the `docs/` folder
-		- Please contact the ODE with qny questions
+		- Please contact the ODE team with any questions
 	6.	Format your code as outlined in the style guide
 
+## Release Processes
+### Quarterly Release
+At the end of each quarter, a 'release/version' branch will be created from the develop branch for all release changes. This branch should be used to stabilize the code and prepare for a release. Any non-release changes continue to be pushed to the develop branch as part of the ongoing development process. Validation testing will be conducted on the release branch to ensure that the code is stable and ready for release. Once the release branch is stable, it will be merged into the master branch and tagged with the release version. If hotfixes are necessary during validation, a hotfix branch is created from the master branch. To conclude the release process, changes from the master branch, including all updates and hotfixes, are merged back into the develop branch.
```
[the remark this answers is not in the document]
Comment (iyourshaw) [reply docs/contributing_guide.md:56]: I agree that cutting hotfixes from the release branch  makes more sense in this case.  Release should only be merged into master after validation testing.

File: docs/contributing_guide.md [hunk 4]
```diff
@@ -50,9 +48,16 @@ All pull requests will be reviewed by the JPO-ODE team. The team member will eit
...
+
+### Hotfix Release
+If a critical bug or issue is found in the master branch, a hotfix branch should be created from the master branch. The hotfix branch should be used to fix the critical bug or issue. Once the hotfix branch is stable, it can be merged into the master branch and tagged with the release version. The master branch will then be merged back into the develop branch to ensure that the develop branch is up to date with the latest changes.
```
[the remark this answers is not in the document]
Comment (iyourshaw) [reply docs/contributing_guide.md:59]: IMO yes the version that the master branch is tagged with should always be kept up to date with the latest release version, including hotfixes.

File: docs/contributing_guide.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmccoystephenson) [reply docs/contributing_guide.md]: The release branch is merged into master after completing validation testing, so yes I agree it makes more sense to cut hotfixes from the release branch. This has been updated.

File: docs/contributing_guide.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dmccoystephenson) [reply docs/contributing_guide.md:61]: Yes, every time a hotfix is released we will tag the master branch with a newer patch version (e.g. 2.1.0 -> 2.1.1)

File: docs/dockerhub.md [hunk 5]
```diff
@@ -28,49 +27,41 @@ Some functionality will be unreachable without the participation of the followin
...
-                                     # file isn't full to ensure that retention can delete or compact old data.
-    depends_on:
-      - zookeeper
-    volumes:
-      - ${DOCKER_SHARED_VOLUME_WINDOWS}/var/run/docker.sock:/var/run/docker.sock
+      KAFKA_CREATE_TOPICS: "topic.OdeBsmPojo:1:1,topic.OdeSpatTxPojo:1:1,topic.OdeSpatPojo:1:1,topic.OdeSpatJson:1:1,topic.FilteredOdeSpatJson:1:1,topic.OdeSpatRxJson:1:1,topic.OdeSpatRxPojo:1:1,topic.OdeBsmJson:1:1,topic.FilteredOdeBsmJson:1:1,topic.OdeTimJson:1:1,topic.OdeTimBroadcastJson:1:1,topic.J2735TimBroadcastJson:1:1,topic.OdeDriverAlertJson:1:1,topic.Asn1DecoderInput:1:1,topic.Asn1DecoderOutput:1:1,topic.Asn1EncoderInput:1:1,topic.Asn1EncoderOutput:1:1,topic.SDWDepositorInput:1:1,topic.OdeTIMCertExpirationTimeJson:1:1,topic.OdeRawEncodedBSMJson:1:1,topic.OdeRawEncodedSPATJson:1:1,topic.OdeRawEncodedTIMJson:1:1,topic.OdeRawEncodedMAPJson:1:1,topic.OdeMapTxPojo:1:1,topic.OdeMapJson:1:1,topic.OdeRawEncodedSSMJson:1:1,topic.OdeSsmPojo:1:1,topic.OdeSsmJson:1:1,topic.OdeRawEncodedSRMJson:1:1,topic.OdeSrmTxPojo:1:1,topic.OdeSrmJson:1:1,topic.OdeRawEncodedPSMJson:1:1,topic.OdePsmTxPojo:1:1,topic.OdePsmJson:1:1"
```
[the remark this answers is not in the document]
Comment (dmccoystephenson) [reply docs/dockerhub.md:59]: Yes! This is out of date. I've updated the example docker-compose.yml file provided in the `dockerhub.md` file. Thank you for catching that.

File: jpo-ode-consumer-example/README.md [hunk 6]
```diff
@@ -49,7 +49,7 @@ The IP used is the location of the Kafka endpoints.
 ####Create, alter, list, and describe topics.
 
 ```
-kafka-topics --zookeeper 192.168.1.151:2181 --list
+kafka-topics --bootstrap-server=192.168.1.151:9092 --list
```
[the remark this answers is not in the document]
Comment (dmccoystephenson) [reply jpo-ode-consumer-example/README.md:52]: This doesn't appear to be necessary & has now been updated.

File: jpo-ode-svcs/src/main/resources/application.properties (hunk 1 above)
Comment (dmccoystephenson) [reply jpo-ode-svcs/src/main/resources/application.properties:22]: SDX! I have updated non-code references (documentation & comments) from 'SDW' to 'SDX'

**Comment by dmccoystephenson** [comment]
This PR is ready for re-review @dan-du-car

**Review by dan-du-car — APPROVED** [review]

**Pull request merged**